### ISSUE TITLE
Error launching LSP server in Neovim with string type initializationOptions

### ISSUE THREAD

**Issue by folliehiyuki** [StyraInc/regal#578]
According to the LSP specs (https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initialize), `initializationOptions` can be of any types. Regal currently forces the value to be string, leading to a json unmarshall error in Neovim.

https://github.com/StyraInc/regal/blob/3205762160dab68f5766d58446cde74617ce9b9d/internal/lsp/messages.go#L24

Neovim sends the request with `initializationOptions` being an empty object if not explicitly specified, hence the error. An example client request:
```bash
{
  "capabilities": {
    "textDocument": {
      "callHierarchy": {
        "dynamicRegistration": false
      },
      "codeAction": {
        "codeActionLiteralSupport": {
          "codeActionKind": {
            "valueSet": [
              "",
              "quickfix",
              "refactor",
              "refactor.extract",
              "refactor.inline",
              "refactor.rewrite",
              "source",
              "source.organizeImports"
            ]
          }
        },
        "dataSupport": true,
        "dynamicRegistration": false,
        "isPreferredSupport": true,
        "resolveSupport": {
          "properties": [
            "edit"
          ]
        }
      },
      "completion": {
        "completionItem": {
          "commitCharactersSupport": true,
          "deprecatedSupport": true,
          "documentationFormat": [
            "markdown",
            "plaintext"
          ],
          "insertReplaceSupport": true,
          "insertTextModeSupport": {
            "valueSet": [
              1,
              2
            ]
          },
          "labelDetailsSupport": true,
          "preselectSupport": true,
          "resolveSupport": {
            "properties": [
              "documentation",
              "detail",
              "additionalTextEdits",
              "sortText",
              "filterText",
              "insertText",
              "textEdit",
              "insertTextFormat",
              "insertTextMode"
            ]
          },
          "snippetSupport": true,
          "tagSupport": {
            "valueSet": [
              1
            ]
          }
        },
        "completionItemKind": {
          "valueSet": [
            1,
            2,
            3,
            4,
            5,
            6,
            7,
            8,
            9,
            10,
            11,
            12,
            13,
            14,
            15,
            16,
            17,
            18,
            19,
            20,
            21,
            22,
            23,
            24,
            25
          ]
        },
        "completionList": {
          "itemDefaults": [
            "commitCharacters",
            "editRange",
            "insertTextFormat",
            "insertTextMode",
            "data"
          ]
        },
        "contextSupport": true,
        "dynamicRegistration": false,
        "insertTextMode": 1
      },
      "declaration": {
        "linkSupport": true
      },
      "definition": {
        "linkSupport": true
      },
      "documentHighlight": {
        "dynamicRegistration": false
      },
      "documentSymbol": {
        "dynamicRegistration": false,
        "hierarchicalDocumentSymbolSupport": true,
        "symbolKind": {
          "valueSet": [
            1,
            2,
            3,
            4,
            5,
            6,
            7,
            8,
            9,
            10,
            11,
            12,
            13,
            14,
            15,
            16,
            17,
            18,
            19,
            20,
            21,
            22,
            23,
            24,
            25,
            26
          ]
        }
      },
      "hover": {
        "contentFormat": [
          "markdown",
          "plaintext"
        ],
        "dynamicRegistration": false
      },
      "implementation": {
        "linkSupport": true
      },
      "publishDiagnostics": {
        "relatedInformation": true,
        "tagSupport": {
          "valueSet": [
            1,
            2
          ]
        }
      },
      "references": {
        "dynamicRegistration": false
      },
      "rename": {
        "dynamicRegistration": false,
        "prepareSupport": true
      },
      "semanticTokens": {
        "augmentsSyntaxTokens": true,
        "dynamicRegistration": false,
        "formats": [
          "relative"
        ],
        "multilineTokenSupport": false,
        "overlappingTokenSupport": true,
        "requests": {
          "full": {
            "delta": true
          },
          "range": false
        },
        "serverCancelSupport": false,
        "tokenModifiers": [
          "declaration",
          "definition",
          "readonly",
          "static",
          "deprecated",
          "abstract",
          "async",
          "modification",
          "documentation",
          "defaultLibrary"
        ],
        "tokenTypes": [
          "namespace",
          "type",
          "class",
          "enum",
          "interface",
          "struct",
          "typeParameter",
          "parameter",
          "variable",
          "property",
          "enumMember",
          "event",
          "function",
          "method",
          "macro",
          "keyword",
          "modifier",
          "comment",
          "string",
          "number",
          "regexp",
          "operator",
          "decorator"
        ]
      },
      "signatureHelp": {
        "dynamicRegistration": false,
        "signatureInformation": {
          "activeParameterSupport": true,
          "documentationFormat": [
            "markdown",
            "plaintext"
          ],
          "parameterInformation": {
            "labelOffsetSupport": true
          }
        }
      },
      "synchronization": {
        "didSave": true,
        "dynamicRegistration": false,
        "willSave": true,
        "willSaveWaitUntil": true
      },
      "typeDefinition": {
        "linkSupport": true
      }
    },
    "window": {
      "showDocument": {
        "support": true
      },
      "showMessage": {
        "messageActionItem": {
          "additionalPropertiesSupport": false
        }
      },
      "workDoneProgress": true
    },
    "workspace": {
      "applyEdit": true,
      "configuration": true,
      "didChangeWatchedFiles": {
        "dynamicRegistration": false,
        "relativePatternSupport": true
      },
      "semanticTokens": {
        "refreshSupport": true
      },
      "symbol": {
        "dynamicRegistration": false,
        "hierarchicalWorkspaceSymbolSupport": true,
        "symbolKind": {
          "valueSet": [
            1,
            2,
            3,
            4,
            5,
            6,
            7,
            8,
            9,
            10,
            11,
            12,
            13,
            14,
            15,
            16,
            17,
            18,
            19,
            20,
            21,
            22,
            23,
            24,
            25,
            26
          ]
        }
      },
      "workspaceEdit": {
        "resourceOperations": [
          "rename",
          "create",
          "delete"
        ]
      },
      "workspaceFolders": true
    }
  },
  "clientInfo": {
    "name": "Neovim",
    "version": "0.9.5"
  },
  "initializationOptions": {},
  "processId": 29478,
  "rootPath": "/home/follie/test-proj",
  "rootUri": "file:///home/follie/test-proj",
  "trace": "off",
  "workspaceFolders": [
    {
      "name": "/home/follie/test-proj",
      "uri": "file:///home/follie/test-proj"
    }
  ]
}
```

**Comment by anderseknert** [comment]
Hey! And thanks for filing this issue. Still very much early days for the Regal language server, and we've only tested it out with VS Code as of yet, and and a first (optional) implementation will land in the OPA VS Code plugin early next week. Our goal was howrver never to limit ourselves to a single client, so we'll do whatever we can to accomodate whatever requirements other clients may have. We just had to start somewhere. 

@charlieegan3 can answer better than me as for the implementation of the LSP here, but I'm not sure I understand this:

> Neovim sends the request with initializationOptions being an empty object if not explicitly specified, leading to a json unmarshall

Why would that be a JSON unmarshal error? Assuming the Neovim client sends "{}" that'd be a valid JSON object, no?

**Comment by charlieegan3** [comment]
Interesting, thanks for filing the issue. We might be able to fix this for now by just dropping that field since don't actually use it yet.

In future, it feels like the correct way to fix this is to build in support for processing these messages from different clients - so this JSON is useful, thanks.

I'll try and do this next week. Thanks for the report @folliehiyuki and do share any updates! 🙂

**Comment by anderseknert** [comment]
Hi @folliehiyuki! And thanks for reporting that. We'll look into it! I believe the first issue reported here has been resolved, so would you mind opening a new issue about this for tracking, and we'll be able to close this one? You can probably just copy-paste the message you just wrote into a new issue.

**Comment by anderseknert** [comment]
Hey @folliehiyuki! It's been a while, and a lot of things have happened in the Regal LSP since then. As I'm not a vim user myself, I'm mainly curious to learn if you've tried any of the new LSP features in Neovim, and how well they work in that context?

**Comment by folliehiyuki** [comment]
I get this error in the latest version of Neovim:

![20240608-121045](https://github.com/StyraInc/regal/assets/67634026/c083ef0c-6be9-4766-9f32-a114f0209041)

Following the stack trace:

![20240608-121246](https://github.com/StyraInc/regal/assets/67634026/c318620e-b84e-492f-8b47-0966ad253a82)

![20240608-121319](https://github.com/StyraInc/regal/assets/67634026/ef04ba50-4e87-48ce-ace2-09c001130465)

It seems like Regal doesn't response with [DocumentDiagnosticReport](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#documentDiagnosticReport) result. [Pull diagnostic](https://microsoft.github.io/language-server-protocol/specification/#textDocument_diagnostic) is implemented in Neovim 0.10.0, according to the changelog (https://neovim.io/doc/user/news-0.10.html#_-new-features).

**Comment by anderseknert** [comment]
Thanks for reporting that! Could you please copy that into a new bug issue?

We do send an empty response for that request, since diagnostics is provided to the client anyway. See the comment in the handler here: https://github.com/StyraInc/regal/blob/29a9c85ceb72f63721c15390138694decfaecf81/internal/lsp/server.go#L1311-L1320

This has been handled well by all the language clients we've tested with so far (VS Code, IntelliJ, Zed, Kate) but apparently not NeoVim. From the code in your screenshot it looks like perhaps it'd work if we sent back `null` rather than an empty struct. Will have to test that! The specification doesn't seem to say what an empty response should look like in this case.

**Comment by anderseknert** [comment]
I've tested sending back `nil` (which will be returned as JSON `null`) now instead of `struct{}{}` , and it seems like that solves the issue you're seeing in Neovim, and AFAICS is handled by all the other editors too. So I'll submit that fix this evening. And since we have a few other bugs either fixed or that will be fixed in the next couple of days, I think we'll do a patch release early next week.

Again, thanks for reporting back! I have now installed Neovim with the LSP plugin too :) And will make sure to test at least that basic things are working for future releases.
